rfcs/20201027-modular-tensorflow-graph-c-api.md

```diff
@@ -0,0 +1,546 @@
+  // set, proper uses default configuration.
+  typedef struct TP_OptimizerConfigFns {
+    size_t struct_size;
+    void* ext;  // reserved for future use
+    TF_Bool (*get_disable_model_pruning)();
+    TF_Bool (*get_implementation_selector)();
+    TF_Bool (*get_function_optimization)();
+    TF_Bool (*get_common_subgraph_elimination)();
+    TF_Bool (*get_arithmetic_optimization)();
+    TF_Bool (*get_debug_stripper)();
+    TF_Bool (*get_constant_folding)();
+    TF_Bool (*get_shape_optimization)();
+    TF_Bool (*get_auto_mixed_precision)();
+    TF_Bool (*get_auto_mixed_precision_mkl)();
+    TF_Bool (*get_pin_to_host_optimization)();
+    TF_Bool (*get_arithmetic_optimization)();
+    TF_Bool (*get_layout_optimizer)();
+    TF_Bool (*get_remapping)();
+    TF_Bool (*get_loop_optimization)();
+    TF_Bool (*get_dependency_optimization)();
+    TF_Bool (*get_memory_optimization)();
+    TF_Bool (*get_auto_parallel)();
+    TF_Bool (*get_scoped_allocator_optimization)();
+  } TP_OptimizerConfigFns;
+
+  #define TP_OPTIMIZER_CONFIG_FNS_STRUCT_SIZE \
+    TF_OFFSET_OF_END(TP_OptimizerConfigFns, get_scoped_allocator_optimization)
+
+  // Struct for Optimizer. Plugin authors must provide an optimize function.
+  // Creation and deletion functions are optional.
+  typedef struct TP_Optimizer {
+    size_t struct_size;
+    void* ext;  // reserved for future use
+    void* (*create_func)();
+    void (*optimize_func)(void*, TF_Buffer*, TF_Buffer*);
+    void (*destory_func)(void*);
+  } TP_Optimizer;
+
+  #define TP_OPTIMIZER_STRUCT_SIZE \
+    TF_OFFSET_OF_END(TP_Optimizer, destory_func)
+
+  typedef struct TP_OptimizerRegistrationParams {
+    size_t struct_size;
+    void* ext;  // reserved for future use
+
+    int32_t major_version;
+    int32_t minor_version;
+    int32_t patch_version;
+
+    // device_type optimizer is registered.
+    const char* device_type;
+    TP_OptimizerConfigFns* config_fns;
+    TP_Optimizer* optimizer;
+  } TP_OptimizerRegistrationParams;
+
+  #define TP_OPTIMIZER_REGISTRARION_PARAMS_STRUCT_SIZE \
+    TF_OFFSET_OF_END(TP_OptimizerRegistrationParams, optimizer)
+
+  void TF_InitGraphPlugin(TP_OptimizerRegistrationParams* params, TF_Status* status);
+
+  #ifdef __cplusplus
+  }  // extern "C"
+  #endif
+  ```
+
+* **Plugin util C API**
+
+  ```cpp
+  #ifdef __cplusplus
+  extern "C" {
+  #endif
+
+  // TF_GrapplerItem represents a combination of a graph, one of more fetch nodes,
+  // and potentially a set of nodes to feed.
+  typedef struct TF_GrapplerItem TF_GrapplerItem;
+
+  // Get TF_GrapplerItem from TF_Buffer.
+  TF_GrapplerItem* TF_GetGrapplerItem(TF_Buffer* buffer);
+
+  // Get a set of node names that must be preserved. This includes feed and
+  // fetch nodes, keep_ops, init_ops. Fills in `max_values` and `storage_size`,
+  // they will be used in `TF_GetNodesToPreserveList`
+  void TF_GetNodesToPreserveSize(TF_GrapplerItem* item, int* max_values,
```

Comment:
Should `max_values` be `num_values`? My guess is this function returns an actual value count.

Reply:
Yes, it is an actual value count. I have changed.